src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -65,6 +72,47 @@ private void AddExceptionDiagnostic(Diagnostic diagnostic)
         }
 
         /// <summary>
+        /// Returns diagnostics produced by diagnostic analyzers from analyzing a single document.
+        /// <param name="model">Semantic model for the document.</param>
+        /// </summary>
+        public async Task<ImmutableArray<Diagnostic>> GetAnalyzerDiagnosticsFromDocumentAsync(SemanticModel model)
+        {
+            SyntaxTree documentTree = model.SyntaxTree;
+            _analyzedSyntaxTrees = _analyzedSyntaxTrees.Add(documentTree);
```

Comment:
Is this thread-safe assignment? You probably need to use an ImmutableInterlocked.CompareExchange and run the subsequent code only if the thread succeeds in adding the tree to the unique trees set. To avoid race conditions, you would also need to ensure that if some thread doesn't succeed in adding the tree (as some other thread succeeded for the same document), then the losing thread doesn't return until the succeeded thread finishes execution. Basically, I am not sure if just maintain a unique set of analyzed trees will suffice. You would need to have an AsyncLazy computation of `GetPartialDiagnosticsAsync` which multiple threads can share.